oil_lang/expr_eval.py

```diff
@@ -125,10 +130,46 @@ def _PyObjToValue(val):
     return value.Str(val)
 
   elif isinstance(val, list):
-    return value.List([_PyObjToValue(elem) for elem in val])
+    return value.List([elem if isinstance(elem, value_t) else _PyObjToValue(elem) for elem in val])
 
   elif isinstance(val, dict):
-      return value.Dict({k: _PyObjToValue(v) for k, v in val.items()})
+      d = NewDict() # type: Dict[str, value_t]
+      for k, v in val.items():
+          if isinstance(v, value_t):
+            d[k] = v
+          else:
+            d[k] = _PyObjToValue(v)
+
+      return value.Dict(d)
+
+  elif isinstance(val, tuple):
+    return value.Tuple([elem if isinstance(elem, value_t) else _PyObjToValue(elem) for elem in val])
+
+  elif isinstance(val, slice):
+      s = value.Slice(None, None, None)
+      if val.start:
+        s.lower = IntParamBox(val.start)
+
+      if val.stop:
+        s.upper = IntParamBox(val.stop)
+
+      if val.step:
+        s.step = IntParamBox(val.step)
+
+      return s
+
+  elif isinstance(val, xrange):
+      r = value.Range(None, None, None)
+      # awkward, but should go away once everything is typed...
+      l = list(val)
```

Comment:
maybe call this `li` or `L`, I think `l` looks too close to `1` which we're also using here :)